File: packages/sdkwork-utils-python/sdkwork_utils/bloom.py
from __future__ import annotations

import hashlib
import math
from dataclasses import dataclass

LN2 = math.log(2)
MIN_BIT_COUNT = 64


@dataclass
class BloomFilter:
    bit_count: int
    hash_count: int
    bits: bytearray


def estimate_bit_count(expected_items: int, false_positive_rate: float) -> int:
    if expected_items <= 0:
        return MIN_BIT_COUNT
    rate = min(0.25, max(0.0001, false_positive_rate))
    size = (-expected_items * math.log(rate)) / (LN2 * LN2)
    return max(MIN_BIT_COUNT, math.ceil(size))


def estimate_hash_count(expected_items: int, bit_count: int) -> int:
    if expected_items <= 0:
        return 1
    count = (bit_count / expected_items) * LN2
    return max(1, round(count))


def create(expected_items: int, false_positive_rate: float) -> BloomFilter:
    bit_count = estimate_bit_count(expected_items, false_positive_rate)
    hash_count = estimate_hash_count(expected_items, bit_count)
    return BloomFilter(bit_count, hash_count, bytearray(math.ceil(bit_count / 8)))


def add(filter: BloomFilter, value: str) -> None:
    for index in _hash_positions(value.encode("utf-8"), filter.bit_count, filter.hash_count):
        byte_index = index // 8
        bit_index = index % 8
        filter.bits[byte_index] |= 1 << bit_index


def might_contain(filter: BloomFilter, value: str) -> bool:
    return all(
        (filter.bits[index // 8] & (1 << (index % 8))) != 0
        for index in _hash_positions(value.encode("utf-8"), filter.bit_count, filter.hash_count)
    )


def _hash_positions(value: bytes, bit_count: int, hash_count: int) -> list[int]:
    digest = hashlib.sha256(value).digest()
    h1 = int.from_bytes(digest[0:4], "big")
    h2 = int.from_bytes(digest[4:8], "big") or 1
    modulus = bit_count
    return [((h1 + index * h2) % modulus) for index in range(hash_count)]

File: packages/sdkwork-utils-rust/src/bloom.rs
use sha2::{Digest, Sha256};
use std::f64::consts::LN_2;

const MIN_BIT_COUNT: usize = 64;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BloomFilter {
    bits: Vec<u8>,
    bit_count: usize,
    hash_count: usize,
}

pub fn estimate_bit_count(expected_items: usize, false_positive_rate: f64) -> usize {
    if expected_items == 0 {
        return MIN_BIT_COUNT;
    }
    let rate = false_positive_rate.clamp(0.000_1, 0.25);
    let size = -(expected_items as f64) * rate.ln() / (LN_2 * LN_2);
    (size.ceil() as usize).max(MIN_BIT_COUNT)
}

pub fn estimate_hash_count(expected_items: usize, bit_count: usize) -> usize {
    if expected_items == 0 {
        return 1;
    }
    let count = (bit_count as f64 / expected_items as f64) * LN_2;
    count.round().max(1.0) as usize
}

pub fn create(expected_items: usize, false_positive_rate: f64) -> BloomFilter {
    let bit_count = estimate_bit_count(expected_items, false_positive_rate);
    let hash_count = estimate_hash_count(expected_items, bit_count);
    let byte_len = bit_count.div_ceil(8);
    BloomFilter {
        bits: vec![0; byte_len],
        bit_count,
        hash_count,
    }
}

pub fn add(filter: &mut BloomFilter, value: &str) {
    for index in hash_positions(value.as_bytes(), filter.bit_count, filter.hash_count) {
        let byte_index = index / 8;
        let bit_index = index % 8;
        filter.bits[byte_index] |= 1 << bit_index;
    }
}

pub fn might_contain(filter: &BloomFilter, value: &str) -> bool {
    hash_positions(value.as_bytes(), filter.bit_count, filter.hash_count)
        .iter()
        .all(|index| {
            let byte_index = index / 8;
            let bit_index = index % 8;
            (filter.bits[byte_index] & (1 << bit_index)) != 0
        })
}

pub fn bit_count(filter: &BloomFilter) -> usize {
    filter.bit_count
}

pub fn hash_count(filter: &BloomFilter) -> usize {
    filter.hash_count
}

fn hash_positions(value: &[u8], bit_count: usize, hash_count: usize) -> Vec<usize> {
    let digest = Sha256::digest(value);
    let h1 = read_u32(&digest[0..4]) as u64;
    let mut h2 = read_u32(&digest[4..8]) as u64;
    if h2 == 0 {
        h2 = 1;
    }

    let modulus = bit_count as u64;
    (0..hash_count)
        .map(|index| ((h1.wrapping_add(index as u64 * h2)) % modulus) as usize)
        .collect()
}

fn read_u32(bytes: &[u8]) -> u32 {
    u32::from_be_bytes([bytes[0], bytes[1], bytes[2], bytes[3]])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bloom_helpers() {
        assert_eq!(estimate_bit_count(128, 0.01), 1227);
        assert_eq!(estimate_hash_count(128, 1227), 7);

        let mut filter = create(128, 0.01);
        add(&mut filter, "hello");
        add(&mut filter, "world");
        assert!(might_contain(&filter, "hello"));
        assert!(might_contain(&filter, "world"));
        assert!(!might_contain(&filter, "__sdkwork_bloom_absent__"));
    }
}
